### PR TITLE
ESM, refactor, options API, example, etc

### PR DESCRIPTION
**Pull request by tbeseda**
- use npm@8
- convert to ESM
- reorganize files
- add eslint
  - with `sourceType = module` config
- use markdown-it-toc-and-anchor 4.2.0
  - the macdonst version isn't currently installable, but we can revert later
- expand options API
  - allow user plugins
  - allow user option overrides for bundled/default plugins
- changed result API
  - not final, but now's the time to do it, so I propose:
    - `html` instead of children, `tocHtml` instead of docOutline, `slug` instead of titleSlug
- update /test
- add /example
- readme

the example.md rendered as HTML:

image

### CODE REVIEW

**Comment by tbeseda** [comment]
usage in arc.codes would look like

```js
const options = {
  hljs: { classString: 'hljs mb0 mb1-lg relative' },
  pluginOverrides: { markdownItClass: classMapping },
}

const result = await render(file, options)
```

then there are a few changes to what arc.codes would expect in `result`, but it's just naming things in a more agnostic/sensible way.

update: a draft PR for arc.codes https://github.com/architect/arc.codes/pull/553

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+---
```
Comment (macdonst) [inline .github/ISSUE_TEMPLATE/bug_report.md]: A cool feature of the https://github.com/architect/.github/tree/main/.github folder is that you don't need to copy these files into the repo. You can but by using https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file we can limit these files to one repo to make updating easier.

File: package.json [hunk 2]
```diff
@@ -1,24 +1,57 @@
...
+    "types/index.d.ts"
+  ],
   "dependencies": {
     "@architect/syntaxes": "github:architect/syntaxes#v1.2.1",
     "highlight.js": "^11.5.1",
     "markdown-it": "^13.0.0",
     "markdown-it-external-anchor": "^1.0.0",
-    "markdown-it-toc-and-anchor": "github:macdonst/markdown-it-toc-and-anchor",
-    "querystring": "^0.2.1",
+    "markdown-it-toc-and-anchor": "^4.2.0",
```
Comment (macdonst) [inline package.json:31]: Oh man, I think we wanna use my fork, `markdown-it-toc-and-anchor` uses `uslug` and it badly impacts coldstart time.

**Review by macdonst — APPROVED** [review]
This looks great to me other than the uslug issue but I'd say merge it then we can keep iterating.

File: package.json (hunk 2 above)
Comment (tbeseda) [reply package.json:31]: Yes I noted above that the fork isn't installable.
Also, does it use uslug even if I pass it the custom slugify function?
I left an item in the Readme's TODO section to possibly use the fork. So I'll merge for now and we can quickly reconsider.

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 1 above)
Comment (tbeseda) [reply .github/ISSUE_TEMPLATE/bug_report.md]: Oh right! thanks for the reminder. I will trim them down before merging. thanks

File: package.json (hunk 2 above)
Comment (macdonst) [reply package.json:31]: That's weird that my fork didn't install for you. Perhaps it needs to be `git+https://github:macdonst/markdown-it-toc-and-anchor.git`?

Yes, even if you pass it a custom slugify function it still does `import uslug from "uslug";` and that import is the problem is `uslug` loads a bunch of arrays of which represent all the unicode characters, IIRC. It's the parsing of the arrays on import during the coldstart which takes forever.

Comment (tbeseda) [reply package.json:31]: It's this command. I believe it needs to be `npm run build`. `yarn` has the build shortcut but npm doesn't
https://github.com/macdonst/markdown-it-toc-and-anchor/blob/master/package.json#L47

That's a bummer about uslug. maybe we can send the original a PR with some benchmarks if we find an equivalent. or with a way to opt out

Comment (macdonst) [reply package.json:31]: I just committed a change to make it `npm run build`

**Pull request merged**